src/db/roomstore.ts

```diff
@@ -0,0 +1,347 @@
+    discord_iconurl?: string|null;
+    discord_iconurl_mxc?: string|null;
+    update_name?: number|boolean|null;
+    update_topic?: number|boolean|null;
+    update_icon?: number|boolean|null;
+    plumbed?: number|boolean|null;
+}
+
+export class RemoteStoreRoom {
+    public data: IRemoteRoomData;
+    constructor(public readonly roomId: string, data: IRemoteRoomData) {
+        for (const k of ["discord_guild", "discord_channel", "discord_name",
+        "discord_topic", "discord_iconurl", "discord_iconurl_mxc", "discord_type"]) {
+            data[k] = typeof(data[k]) === "number" ? String(data[k]) : data[k] || null;
+        }
+        for (const k of ["update_name", "update_topic", "update_icon", "plumbed"]) {
+            data[k] = Number(data[k]) || 0;
+        }
+        this.data = data;
+    }
+
+    public getId() {
+        return this.roomId;
+    }
+
+    public get(key: string): string|boolean|null {
+        return this.data[key];
+    }
+
+    public set(key: string, value: string|boolean|null) {
+        this.data[key] = typeof(value) === "boolean" ? Number(value) : value;
+    }
+}
+
+export class MatrixStoreRoom {
+    constructor(public readonly roomId: string) { }
+
+    public getId() {
+        return this.roomId;
+    }
+}
+
+export interface IRoomStoreEntry {
+    id: string;
+    matrix: MatrixStoreRoom|null;
+    remote: RemoteStoreRoom|null;
+}
+
+const ENTRY_CACHE_LIMETIME = 30000;
+
+// XXX: This only implements functions used in the bridge at the moment.
+export class DbRoomStore {
+
+    private entriesMatrixIdCache: Map<string, {e: IRoomStoreEntry[], ts: number}>;
+
+    constructor(private db: IDatabaseConnector) {
+        this.entriesMatrixIdCache = new Map();
+    }
+
+    public async upsertEntry(entry: IRoomStoreEntry) {
+        const promises: Promise<void>[] = [];
+
+        const row = (await this.db.Get("SELECT * FROM room_entries WHERE id = $id", {id: entry.id})) || {};
+
+        if (!row.id) {
+            // Doesn't exist at all, create the room_entries row.
+            const values = {
+                id: entry.id,
+                matrix: entry.matrix ? entry.matrix.roomId : null,
+                remote: entry.remote ? entry.remote.roomId : null,
+            };
+            try {
+                await this.db.Run(`INSERT INTO room_entries VALUES ($id, $matrix, $remote)`, values);
+                log.verbose("Created new entry " + entry.id);
+            } catch (ex) {
+                log.error("Failed to insert room entry", ex);
+                throw Error("Failed to insert room entry");
+            }
+        }
+
+        const matrixId = entry.matrix ? entry.matrix.roomId : null;
+        const remoteId = entry.remote ? entry.remote.roomId : null;
+        const mxIdDifferent = matrixId !== row.matrix_id;
+        const rmIdDifferent = remoteId !== row.remote_id;
+        // Did the room ids change?
+        if (row.id && (mxIdDifferent || rmIdDifferent)) {
```

Comment:
yes, we check again for `row.id` here, but it is kinda....redundant, we do the same check twice

Reply:
oh derp